python/ql/src/experimental/semmle/python/frameworks/Invoke.qll

```diff
@@ -69,15 +69,73 @@ private module Invoke {
   }
 
   /** Provides models for the `invoke` module. */
-  module invoke { }
+  module invoke {
+    /** Gets a reference to the `invoke.context` module. */
+    DataFlow::Node context() { result = invoke_attr("context") }
+
+    /** Provides models for the `invoke.context` module */
+    module context {
+      /** Provides models for the `invoke.context.Context` class */
+      module Context {
+        /** Gets a reference to the `invoke.context.Context` class. */
+        private DataFlow::Node class_(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importMember("invoke.context", "Context")
+          or
+          t.startInAttr("Context") and
+          result = invoke::context()
+          or
+          t.start() and
+          result = invoke_attr("Context")
+          or
+          exists(DataFlow::TypeTracker t2 | result = class_(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `invoke.context.Context` class. */
+        DataFlow::Node class_() { result = class_(DataFlow::TypeTracker::end()) }
+
+        /** Gets a reference to an instance of `invoke.context.Context`. */
+        private DataFlow::Node instance(DataFlow::TypeTracker t) {
+          t.start() and
+          result.asCfgNode().(CallNode).getFunction() =
+            invoke::context::Context::class_().asCfgNode()
+          or
+          exists(DataFlow::TypeTracker t2 | result = instance(t2).track(t2, t))
+        }
+
+        /** Gets a reference to an instance of `invoke.context.Context`. */
+        DataFlow::Node instance() { result = instance(DataFlow::TypeTracker::end()) }
+
+        /** Gets a reference to the `run` or `sudo` methods on a `invoke.context.Context` instance. */
+        private DataFlow::Node instanceRunMethods(DataFlow::TypeTracker t) {
+          t.startInAttr(["run", "sudo"]) and
+          result = invoke::context::Context::instance()
```

Comment:
```suggestion
          result = instance()
```
since we are still inside both modules, right?

Reply:
we could for sure. Do you have a preference for doing so? -- Personally I don't have a big preference for either.

Reply:
I am torn...on the one hand, the long form reads kinda nice, on the other, it feels weird to use the long form inside the module, and initially confused me think we had somehow jumped out..no strong preference, really.